### PR TITLE
[Response to Reviewers] Add Silhouette analysis

### PR DESCRIPTION
**Pull request by gwaybio**
This PR is in response to the following reviewer comment:

> You state “By eye, CellProfiler features demonstrated the most heterogeneity…”, is there perhaps a way of quantifying this? For example, some kind of neighbourhood analysis.

We think this is a good idea, and therefore performed the following analysis:

1. Per feature space, calculate Silhouette score per phenotype in an all vs. one comparison (e.g., All anaphase cells vs. all other cells)
2. Apply PCA to make dimensions of input features consistent (`n_components=50`)
3. Calculate average Silhouette width (per all-vs.-one phenotype and per feature space)

We interpret the Silhouette scores  how well cells of a given phenotype are clustered compared to other cells of the same phenotype. A positive score means cells of the same phenotype are more similar to other cells of the same phenotype (on average) compared to all other cells. A score of 1 indicates complete separation of similar phenotypes from other phenotypes.

## New supplementary figure

supplementary_silhouette_scores

### CODE REVIEW

File: 3.evaluate_model/scripts/nbconverted/calculate_silhouette_metrics.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+import sys
+
+sys.path.append("../utils")
+from split_utils import get_features_data
+
+
+# In[2]:
+
+
+np.random.seed(1234)
```
Comment (jenna-tomkinson) [inline 3.evaluate_model/scripts/nbconverted/calculate_silhouette_metrics.py:28]: Recommend random seed `0` to be consistent with Way Lab standard in other projects.

File: 3.evaluate_model/scripts/nbconverted/calculate_silhouette_metrics.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+# For consistent Silhouette input space dimensionality
+n_pca_components = 50
```
Comment (jenna-tomkinson) [inline 3.evaluate_model/scripts/nbconverted/calculate_silhouette_metrics.py:31]: Where does `50` come from? I know in UMAP we do `2` components, what is the difference when is comes to PCA?

Comment (jenna-tomkinson) [reply 3.evaluate_model/scripts/nbconverted/calculate_silhouette_metrics.py:31]: Also heads up, in the PR comment you say the number of components is `40` but in here it is `50`, recommend confirming which one is correct/most appropriate.

File: 3.evaluate_model/scripts/nbconverted/calculate_silhouette_metrics.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+
+eval_path = pathlib.Path("evaluations")
+
+output_silhouette_results = pathlib.Path(
+    eval_path, "silhouette_score_results.tsv"
+)
+output_silhouette_samples_results = pathlib.Path(
+    eval_path, "silhouette_score_results_per_sample.tsv"
+)
```
Comment (jenna-tomkinson) [inline 3.evaluate_model/scripts/nbconverted/calculate_silhouette_metrics.py]: Recommend making these compressed TSVs for saving space plus that might be the standard convention in this repo.

**Review by jenna-tomkinson — APPROVED** [review]
LGTM! Really nice, clean, and simple PR!

Lookis like this analysis does show that CellProfiler does have the most heterogeneity since it has 6 phenotypes with the top positive silhouette score. Very interesting results!

File: 3.evaluate_model/scripts/nbconverted/calculate_silhouette_metrics.py (hunk 1 above)
Comment (gwaybio) [reply 3.evaluate_model/scripts/nbconverted/calculate_silhouette_metrics.py:28]: i will stick with 1234

File: 3.evaluate_model/scripts/nbconverted/calculate_silhouette_metrics.py (hunk 2 above)
Comment (gwaybio) [reply 3.evaluate_model/scripts/nbconverted/calculate_silhouette_metrics.py:31]: ah! Great catch.

In my experience, 50 is more than enough to capture the majority of the variance in the dataset, which is what we're aiming for. It's more or less an arbitrary number

File: 3.evaluate_model/scripts/nbconverted/calculate_silhouette_metrics.py (hunk 3 above)
Comment (gwaybio) [reply 3.evaluate_model/scripts/nbconverted/calculate_silhouette_metrics.py]: they are super tiny, will stick with tsv so they can be rendered on github

**Comment by gwaybio** [comment]
Thanks for the review @jenna-tomkinson - I caught a couple things too, which I addressed in the recent commits. Merging now!

**Pull request merged**